### PR TITLE
Bug 1742448 - Generate build date at invocation time

### PR DESCRIPTION
**Pull request by badboy**
Implementation of the build date field in `build info`.
Will require Glean updates as well.

---

~~Draft until proposal is formally accepted.~~

### CODE REVIEW

File: glean_parser/swift.py [hunk 1]
```diff
@@ -157,6 +184,9 @@ def output_swift(
         - namespace: The namespace to generate metrics in
         - glean_namespace: The namespace to import Glean from
         - allow_reserved: When True, this is a Glean-internal build
+        - with_buildinfo: If "true" the `GleanBuildInfo` is generated.
+          Otherwise generation of that file is skipped.
+          Defaults to "true".
```
Comment (brizental) [inline glean_parser/swift.py:189]: Missing the actual `build_info` param documentation.

Comment (brizental) [reply glean_parser/swift.py:189]: Kotlin is missing both params documentation.

**Review by brizental — CHANGES_REQUESTED** [review]
Would you mind filing follow up bugs to implement this for other languages?

File: glean_parser/swift.py (hunk 1 above)
Comment (badboy) [reply glean_parser/swift.py:189]: Oh damn, will fill it in.

File: glean_parser/swift.py [hunk 2]
```diff
@@ -134,6 +134,33 @@ def variable_name(var: str) -> str:
...
+def generate_build_date(date: Optional[str]) -> str:
+    """
+    Generate the build timestamp.
+    """
+
+    ts = util.build_date(date)
+
+    data = [
+        ("year", ts.year),
+        ("month", ts.month),
```
[the remark this answers is not in the document]
Comment (badboy) [reply glean_parser/swift.py:151]: Nope.

File: glean_parser/util.py [hunk 3]
```diff
@@ -407,6 +421,31 @@ def validate_expires(expires: str) -> None:
         )
 
 
+def build_date(date: Optional[str]) -> datetime.datetime:
+    """
+    Generate the build timestamp.
+
+    If `date` is set to `0` a static unix epoch time will be used.
```
[the remark this answers is not in the document]
Comment (badboy) [reply glean_parser/util.py:428]: Easier way to guarantee reproducible builds, though it's basically the same as saying `build_date=1970-01-01T00:00:00`, so we could also leave it out.

**Comment by badboy** [comment]
* [Python](https://bugzilla.mozilla.org/show_bug.cgi?id=1749493)
* [Rust](https://bugzilla.mozilla.org/show_bug.cgi?id=1749495)
* [Glean.js](https://bugzilla.mozilla.org/show_bug.cgi?id=1749494)

File: glean_parser/util.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (brizental) [reply glean_parser/util.py:428]: I'm fine leaving it. I just think it's a bit weird, but harmless.

**Review by brizental — APPROVED** [review]

**Pull request merged**